pallets/dispenser/src/tests.rs

```diff
@@ -0,0 +1,122 @@
+// Polimec Blockchain – https://www.polimec.org/
+// Copyright (C) Polimec 2022. All rights reserved.
+
+// The Polimec Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Polimec Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@polimec.org
+
+use super::*;
+use crate as pallet_dispenser;
+use crate::mock::*;
+use frame_support::{assert_noop, assert_ok};
+use polimec_common::credentials::InvestorType;
+use polimec_common_test_utils::{generate_did_from_account, get_mock_jwt};
+use sp_runtime::DispatchError;
+
+mod admin {
+	use super::*;
+	#[test]
+	fn initial_amount_is_correctly_set() {
+		ExtBuilder::default().build().execute_with(|| {
+			// Test that the initial dispense amount is set to the InitialDispenseAmount constant.
+			assert_eq!(DispenseAmount::<Test>::get(), <Test as pallet_dispenser::Config>::InitialDispenseAmount::get());
+		});
+	}
+
+	/// Test that only the Admin can change the dispense amount.
+	#[test]
+	fn only_admin_can_change_dispense_amount() {
+		ExtBuilder::default().build().execute_with(|| {
+			assert_noop!(Dispenser::set_dispense_amount(RuntimeOrigin::signed(1), 100), DispatchError::BadOrigin);
+			assert_ok!(Dispenser::set_dispense_amount(RuntimeOrigin::signed(Admin::get()), 1000));
+			assert_eq!(DispenseAmount::<Test>::get(), 1000);
+		});
+	}
+
+	#[test]
+	fn amount_has_to_be_higher_then_free_amount() {
+		ExtBuilder::default().build().execute_with(|| {
+			assert_noop!(
+				Dispenser::set_dispense_amount(
+					RuntimeOrigin::signed(Admin::get()),
+					<Test as pallet_dispenser::Config>::FreeDispenseAmount::get()
+				),
+				Error::<Test>::DispenseAmountTooLow
+			);
+		});
+	}
+}
+
+mod dispense {
+	use super::*;
+	#[test]
+	fn user_can_dispense_for_free() {
```

Comment:
The mock runtime doesn't charge a fee for executing extrinsics. I assume this is to make pallet tests easier since a lot of pallets modify your balance outside of the fees charged.
You can double-check this with this simple test:
```rust
	#[test]
	fn sandbox() {
		ExtBuilder::default().build().execute_with(|| {
			assert_eq!(Balances::free_balance(1), 0);
			assert_ok!(System::remark(RuntimeOrigin::signed(1), vec![69, 69]));
		});
	}

```

I think you need to modify the mock runtime to actually charge fees (if that is even possible)

Reply:
Why exactly? We are testing the correct execution of the logic within the extrinsics. For the "free execution" tests, we have an integration test that checks the validate and pre_dispatch execution. And I ran the extrinsic on zombienet to see if execution was free, and this worked. So I thing these three together should show the correct working of the pallet.

Reply:
the test is called `user_can_dispense_for_free` but you are not testing that because there are no fees in this mock runtime, so if you think this is tested in other places like integration-tests, then I would remove this test and maybe leave a comment where this test can be found.

I see though that we have 2 dispenser tests called `dispenser_signed_extensions_pass_for_new_account` and `dispenser_works_with_runtime_values` But I don't see it testing that a user with 0 balance can call dispense, while other calls would require fees/ED.